pkg/pgxconn/pgx_conn.go

```diff
@@ -19,35 +23,97 @@ type PgxBatch interface {
 type PgxConn interface {
 	Close()
 	Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error)
-	Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error)
+	Query(ctx context.Context, sql string, args ...interface{}) (PgxRows, error)
 	QueryRow(ctx context.Context, sql string, args ...interface{}) pgx.Row
 	CopyFrom(ctx context.Context, tableName pgx.Identifier, columnNames []string, rowSrc pgx.CopyFromSource) (int64, error)
 	CopyFromRows(rows [][]interface{}) pgx.CopyFromSource
 	NewBatch() PgxBatch
 	SendBatch(ctx context.Context, b PgxBatch) (pgx.BatchResults, error)
 }
 
+type PgxRows interface {
+	Next() bool
+	Scan(...interface{}) error
+	Err() error
+	Close()
+}
+
+// This is only used by querier to log the time consumed
+// by SQL queries executed using Query(), QueryRow()
+func NewQueryLoggingPgxConn(pool *pgxpool.Pool) PgxConn {
+	return &loggingConnImpl{
+		connImpl: connImpl{Conn: pool},
+		Conn:     pool,
+	}
+}
+
 func NewPgxConn(pool *pgxpool.Pool) PgxConn {
 	return &connImpl{
 		Conn: pool,
 	}
 }
 
+type loggingConnImpl struct {
+	connImpl
+	Conn *pgxpool.Pool
+}
+
 type connImpl struct {
 	Conn *pgxpool.Pool
 }
 
+type pgxRows struct {
+	pgx.Rows
+	sqlQuery  string
+	args      []interface{}
+	startTime time.Time
+	logged    bool
+}
+
+func (p *loggingConnImpl) Query(ctx context.Context, sql string, args ...interface{}) (PgxRows, error) {
+	startTime := time.Now()
+	rows, err := p.Conn.Query(ctx, sql, args...)
+	return &pgxRows{Rows: rows, sqlQuery: sql, args: args, startTime: startTime}, err
+}
+
+func (p *loggingConnImpl) QueryRow(ctx context.Context, sql string, args ...interface{}) pgx.Row {
+	defer logQueryStats(sql, time.Time{}, args...)()
+	return p.Conn.QueryRow(ctx, sql, args...)
+}
+
 func (p *connImpl) Close() {
 	conn := p.Conn
 	p.Conn = nil
 	conn.Close()
 }
 
-func (p *connImpl) Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error) {
-	return p.Conn.Exec(ctx, sql, arguments...)
+func (p *pgxRows) Next() bool {
```

Comment:
nit adding functions from the same class in the same section of the file would be better. right now we have connImpl methods interspersed with pgxRows and loggingConnImpl